### DISCUSSION TITLE
Custom authStore and TypeScript

### DISCUSSION THREAD

**Discussion by ollema** [pocketbase/pocketbase#4300] [Q&A] [answered]
Hi!

I'm working on an auth store that integrates nicely with [`@capacitor/preferences`](https://capacitorjs.com/docs/apis/preferences).

```ts
import { BaseAuthStore, type AuthModel } from 'pocketbase';
import { Preferences } from '@capacitor/preferences';

const authKey = 'pb_auth';

class PreferencesAuthStore extends BaseAuthStore {
	private queue: Array<() => Promise<void>> = [];

	constructor() {
		super();
	}

	save(token: string, model?: AuthModel): void {
		super.save(token, model);
		this._enqueue(
			async () => await Preferences.set({ key: authKey, value: JSON.stringify({ token, model }) })
		);
	}

	clear(): void {
		super.clear();
		this._enqueue(async () => await Preferences.remove({ key: authKey }));
	}

	async loadInitial() {
		const result = await Preferences.get({ key: authKey });
		const parsed = JSON.parse(result.value || '{}') || {};
		this.save(parsed.token || '', parsed.model || null);
	}

	private _enqueue(asyncCallback: () => Promise<void>) {
		this.queue.push(asyncCallback);

		if (this.queue.length == 1) {
			this._dequeue();
		}
	}

	private _dequeue() {
		if (!this.queue.length) {
			return;
		}

		this.queue[0]().finally(() => {
			this.queue.shift();

			if (!this.queue.length) {
				return;
			}

			this._dequeue();
		});
	}
}
```

as you can tell, it is very similar to `AsyncAuthStore` but I need to be able to await the `loadInitial` function in `+layout.ts`:
```ts
export const load = async ({ fetch, url }) => {
	await pb.authStore.loadInitial();
	...
}
```
but then I get:
```
Property 'loadInitial' does not exist on type 'BaseAuthStore'.
```

Is it possible to type the `PocketBase` instance somehow so that TypeScript is aware that this method does indeed exist on the authStore?

**Comment by ganigeorgiev** [comment]
There is no generic option for the `authStore`. I don't fully understand your use case but you could try:

- Using type assertion (`(pb.authStore as yourType).*`)
- Provide a definition for the entire `PocketBase` client object. This could be done like in the second example in [Specify TypeScript definitions](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#specify-typescript-definitions). In the example we use type assertion for the `TypedPocketBase` interface but you should be able to do the same with `declare class`:
```typescript
import PocketBase from "pocketbase";

declare class PocketBase {
    authStore: yourType
    // ...
}

const pb = new PocketBase("...")
```

Alternatively, you can also move and export `loadInitial` as a separate helper function and for example call it with `pb.authStore` as argument.

If you just want to be able to access the store's `loadInitial` method, you can also try with the brackets notation: `pb.authStore['loadInitial']()`.

**Reply by ollema** [reply]
thanks @ganigeorgiev!

I am already using `TypedPocketBase` so I will try to adapt that so that it also includes a override for the authStore.

---

> Alternatively, you can also move and export loadInitial as a separate helper function and for example call it with pb.authStore as argument.

how could I do that?

to be more specific, how can I add a method that await the initial value being set in the `AsyncAuthStore`?

normally, when calling `_loadInitial()` or `save()` in the original `AsyncAuthStore`, the operation is only enqueued. I could not figure out how I can make sure that the operation has finished.

my use case is that when I initialise the PocketBase client I load the initial value from an external async store.

I need this value to be read before loading other data, since that data might be dependant on whether or not someone is logged in or not.

therefore, I need a way of ensuring that the data is loaded, which is the `loadInitial` method in my custom auth store above

**Reply by ganigeorgiev** [reply, marked as answer]
I'm confused and without more details about your use case I really don't think I can help you further.

> my use case is that when I initialise the PocketBase client I load the initial value from an external async store.

Why can't you await the `Preferences` before initializing the client and then load it with `pb.authStore.save()`?

`pb.authStore.token`/`pb.authstore.model` will be read from the fields that are kept as part of the store object (notice the `super.save/clear` calls) and it shouldn't matter when the `Preferences` will be updated (the `Preferences` in your case is just there for persistence so that in case your app restarts can continue from where it left).

**Reply by ollema** [reply]
ah, you are totally right. I was doing things backwards.

this works well it seems:

`// src/lib/pocketbase.ts`
```ts
import PocketBase, { AsyncAuthStore } from 'pocketbase';
import type { TypedPocketBase } from '$lib/pocketbase-types';
import { Preferences } from '@capacitor/preferences';

const authKey = 'pb_auth';

const preferencesAsyncAuthStore = new AsyncAuthStore({
	save: async (serialized) => await Preferences.set({ key: authKey, value: serialized }),
	clear: async () => await Preferences.remove({ key: authKey })
});

export const pb = new PocketBase(
	'https://example.com',
	preferencesAsyncAuthStore
) as TypedPocketBase;

export async function loadInitial() {
	const serialized = (await Preferences.get({ key: authKey })).value;
	const parsed = JSON.parse(serialized || '{}') || {};
	if (serialized) {
		pb.authStore.save(parsed.token || '', parsed.model || null);
	}
}
```

`// src/routes/+layout.ts`
```ts
import { loadInitial } from '$lib/pocketbase';

export const load = async ({ fetch }) => {
	await loadInitial();
}
```
